### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by chrisns**

### CODE REVIEW

**Comment by chrisns** [comment]
@techjacker @shabe @danielepolencic thoughts welcome while I try and put this together

File: principles/testing.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+Knowing things work is really important, it allows us to iterate and change confidently without unintentionally breaking things.
+
+## Manual testing
+...is **NOT** the solution (no more ice cream cones)
+
+![Ice cream cone testing antipatern](https://techblog.workiva.com/sites/techblog/files/tech_blogs/engineering-quality-test-ice-cream-cone-tech-blog-k7776-20151019.png)
+
+
+## Automated testing
+Writing tests requires testable code, which is why its always easier to write the test first.
```
Comment (techjacker) [inline principles/testing.md]: typo: "...which is why **_it's_** always"

File: principles/testing.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+*For an object orientated language this would be tests for each method in a class.*
+
+#### Component
+Components are the next step up the pyramid from the individual units and they allow you to begin to test how a specfic component works as a whole, where dependencies are still mocked.
+
+*For an object orientated language this would be tests for the class as a whole.*
+
+#### Integration
+This step allows you to test how your components work together, where you can still
```
Comment (techjacker) [inline principles/testing.md]: "where you can still...?"

File: principles/testing.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+Small services where the code is probaly only hundreds of lines long shouldn't have the same level of complexity as a monolith and therefore testing is structured slightly differently.
+
+### Units
+Nothing different here, everything as above applies.
+
+### Integration
+This level of testing covers a number of system components working together, however external services are all still mocked out. You might exercise this level of testing by sending a request, check the service layer has done its job and then validate the response back.
+
+### Contracts
+The fundemental microservice architecture depends on services communicating with each other. Whilst the internal implmentation of a service is independant and at the discression of the team responsible for developing it; interface and outward facing API **must remain consistent**.
```
Comment (techjacker) [inline principles/testing.md]: "service is **independent**"

File: principles/testing.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+We commonly use [JSON Schema](http://json-schema.org/) or [XML Schema](https://www.w3.org/standards/xml/schema) for defining these contract, it is often helpful to your consumers to serve a copy of the current schema from the service itself.
+
+#### Breaking changes
+You should only be able to introduce a breaking change (one that may/will break consumer integration with your service) with a contract change, it's therefore really important to communicate these, giving your consumers adequte time and notice to update. Not all contract changes need to be breaking however and backwards compatability is highly encouraged.
+
+### End-to-end
```
Comment (techjacker) [inline principles/testing.md]: I think this section is superfluous.

File: principles/testing.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+## Microservices
+Small services where the code is probaly only hundreds of lines long shouldn't have the same level of complexity as a monolith and therefore testing is structured slightly differently.
+
+### Units
+Nothing different here, everything as above applies.
+
+### Integration
+This level of testing covers a number of system components working together, however external services are all still mocked out. You might exercise this level of testing by sending a request, check the service layer has done its job and then validate the response back.
+
+### Contracts
```
Comment (techjacker) [inline principles/testing.md]: I think we should include some info on [consumer driven contract tests](https://www.thoughtworks.com/radar/tools/pact-pacto) here and create a new section entitled "API Validation tools" for this info on JSONSchema.

Comment (chrisns) [inline principles/testing.md]: Might be worth spinning this off in to a separate topic aside of testing maybe, thoughts?

**Comment by danielepolencic** [comment]
I just realised that there's a similar repo created by Jonny: https://github.com/UKHomeOffice/testing-guide-for-developers

Do you need to keep both: this PR and Jonny's repo?

**Comment by chrisns** [comment]
@danielepolencic already had this conversation with @timgent the original plan was for this PR to supercede that repo, which is why it's taken some inspiration/basis from it

File: principles/testing.md (hunk 5 above)
Comment (chrisns) [inline principles/testing.md]: https://github.com/UKHomeOffice/engineering-and-architecture-board/issues/243

File: principles/testing.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+## Performance
+Many of your performance testing needs can be answered by tooling your other test layers with instrumentation to monitor how long actions take, often you'll have overarching targets, perhaps an SLA (Service level agreement) where you're required to make a commitment along the lines of handling `n` simultaneous requests, you should look to cater for this in your end-to-end tests and test it in the pipeline.
+
+It is generally considered acceptable to test a `3-5x` multiplier of what is considered 'a peak level of traffic'. Your service manager/designer should be able to provide guidance on what your service should consider normal, peak and what it should be built to be handle.
+
+When your application is deployed on to a shared infrastructure such as `DSP` it is particularly important to be clear about what you're testing when you're doing performance testing, so as you're not unnecessarily testing the infrastructure of which you have little or no control over.
```
Comment (danielepolencic) [inline principles/testing.md]: DSP -> ACP

**Pull request merged**